### PR TITLE
mgr/cephadm: Make daemon actions asynchronous

### PR DESCRIPTION
**Pull request by sebastian-philipp**
Fixes broken 

* `ceph orch redeploy <current active mgr>`
* `ceph orch redeploy mon...`

Replaces #36720

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2020-08-24_15:45:33-rados:cephadm-wip-swagner-testing-2020-08-24-1245-distro-basic-smithi/5372897/

**Comment by neha-ojha** [comment]
@sebastian-philipp want to me this against my cephadm based upgrade suite?

**Comment by sebastian-philipp** [comment]
not yet. there is a test failure that needs fixing

**Comment by sebastian-philipp** [comment]
green: https://pulpito.ceph.com/swagner-2020-08-25_16:35:43-rados:cephadm-wip-swagner-testing-2020-08-25-1535-distro-basic-smithi/

**Comment by sebastian-philipp** [comment]
DNM, needs squashing. 

@neha-ojha this PR is good now.

**Comment by neha-ojha** [comment]
> DNM, needs squashing.
> 
> @neha-ojha this PR is good now.

great, I will report back results here.

**Comment by neha-ojha** [comment]
> > DNM, needs squashing.
> > @neha-ojha this PR is good now.
> 
> great, I will report back results here.

@sebastian-philipp I ran your build against my test branch and it is failing with the same error https://pulpito.ceph.com/nojha-2020-08-25_17:55:27-rados:cephadm:partial-upgrade-wip-swagner-testing-2020-08-25-1535-distro-basic-smithi/

**Comment by sebastian-philipp** [comment]
> > > DNM, needs squashing.
> > > @neha-ojha this PR is good now.
> > 
> > 
> > great, I will report back results here.
> 
> @sebastian-philipp I ran your build against my test branch and it is failing with the same error https://pulpito.ceph.com/nojha-2020-08-25_17:55:27-rados:cephadm:partial-upgrade-wip-swagner-testing-2020-08-25-1535-distro-basic-smithi/

Hm this can't work. We're trying to run `orch daemon redeploy --image` on a MGR that is still on 15.2.  

I think we have to:

1. Merge this PR
2. backport to octopus
3. set the upgrade start to the merge commit of the backport

**Comment by neha-ojha** [comment]
> > > > DNM, needs squashing.
> > > > @neha-ojha this PR is good now.
> > > 
> > > 
> > > great, I will report back results here.
> > 
> > 
> > @sebastian-philipp I ran your build against my test branch and it is failing with the same error https://pulpito.ceph.com/nojha-2020-08-25_17:55:27-rados:cephadm:partial-upgrade-wip-swagner-testing-2020-08-25-1535-distro-basic-smithi/
> 
> Hm this can't work. We're trying to run `orch daemon redeploy --image` on a MGR that is still on 15.2.0 .
> 
> I think we have to:
> 
> 1. Merge this PR
> 2. backport to octopus
> 3. set the upgrade start to the merge commit of the backport

@sebastian-philipp makes sense, let's do it then!

**Comment by sebastian-philipp** [comment]
green: https://pulpito.ceph.com/swagner-2020-08-27_07:38:41-rados:cephadm-wip-swagner2-testing-2020-08-26-1329-distro-basic-smithi/

**Comment by adk3798** [comment]
Is this intended to fix redeploying the whole mgr service as well or just the single active mgr? Works perfectly for me with a `ceph orch daemon redeploy <active mgr>` which was broken before but trying `ceph orch redeploy mgr` (with 3 mgrs in this case) left me in a loop where it just keeps failing between the three mgrs forever: https://pastebin.com/G0wYVtQH

**Comment by sebastian-philipp** [comment]
> Is this intended to fix redeploying the whole mgr service as well or just the single active mgr? Works perfectly for me with a `ceph orch daemon redeploy <active mgr>` which was broken before but trying `ceph orch redeploy mgr` (with 3 mgrs in this case) left me in a loop where it just keeps failing between the three mgrs forever: https://pastebin.com/G0wYVtQH

indeed this PR **only** fixes `ceph orch daemon redeploy`. Fixing `ceph orch redeploy` is tracked in https://tracker.ceph.com/issues/45465 

I really don't want to blow up this PR anymore

File: src/pybind/mgr/cephadm/upgrade.py [hunk 1]
```diff
@@ -163,7 +163,7 @@ def _wait_for_ok_to_stop(self, s: DaemonDescription) -> bool:
             if not r.retval:
                 logger.info(f'Upgrade: {r.stdout}')
                 return True
-            logger.error('Upgrade: {r.stderr}')
+            logger.error(f'Upgrade: {r.stderr}')
```
Comment (mgfritch) [inline src/pybind/mgr/cephadm/upgrade.py:166]: :+1:

File: src/pybind/mgr/cephadm/inventory.py [hunk 2]
```diff
@@ -508,6 +515,34 @@ def daemon_cache_filled(self):
...
                    for h in self.get_hosts())
 
+    def schedule_daemon_action(self, host: str, daemon_name: str, action: str):
+        priorities = {
+            'start': 1,
+            'restart': 2,
+            'reconfig': 3,
+            'redeploy': 4,
+            'stop': 5,
+        }
```
Comment (mgfritch) [inline src/pybind/mgr/cephadm/inventory.py:525]: minor, but I wonder if at some point we should refactor these to be on the `DaemonDescription` and track the scheduled action there ...

File: src/pybind/mgr/cephadm/module.py [hunk 3]
```diff
@@ -1659,13 +1648,50 @@ def _daemon_action(self, daemon_type, daemon_id, host, action, image=None):
...
     def daemon_action(self, action: str, daemon_name: str, image: Optional[str]=None) -> str:
         d = self.cache.get_daemon(daemon_name)
 
-        self.log.info(f'{action} daemon {daemon_name}')
-        return self._daemon_action(d.daemon_type, d.daemon_id,
-                                 d.hostname, action, image=image)
+        if action == 'redeploy' and self.daemon_is_self(d.daemon_type, d.daemon_id) \
+                and not self.mgr_service.mgr_map_has_standby():
+            raise OrchestratorError(
+                f'Unable to schedule redeploy for {daemon_name}: No standby MGRs')
```
Comment (adk3798) [inline src/pybind/mgr/cephadm/module.py:1675]: Do we need to have the same check for a standby mgr in `daemon_action`, `_schedule_daemon_action` and the MgrService `fail_over` function? Do they all have circumstances where they're necessary? Other than that LGTM and works locally.

**Comment by sebastian-philipp** [comment]
green: https://pulpito.ceph.com/swagner-2020-08-28_15:31:32-rados:cephadm-wip-swagner-testing-2020-08-28-1411-distro-basic-smithi/

File: src/pybind/mgr/cephadm/module.py (hunk 3 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/module.py:1675]: nope. this is special to MGRs. We just can't redeploy our own MGR.

**Comment by neha-ojha** [comment]
> > > > > DNM, needs squashing.
> > > > > @neha-ojha this PR is good now.
> > > > 
> > > > 
> > > > great, I will report back results here.
> > > 
> > > 
> > > @sebastian-philipp I ran your build against my test branch and it is failing with the same error https://pulpito.ceph.com/nojha-2020-08-25_17:55:27-rados:cephadm:partial-upgrade-wip-swagner-testing-2020-08-25-1535-distro-basic-smithi/
> > 
> > 
> > Hm this can't work. We're trying to run `orch daemon redeploy --image` on a MGR that is still on 15.2.0 .
> > I think we have to:
> > 
> > 1. Merge this PR
> > 2. backport to octopus
> > 3. set the upgrade start to the merge commit of the backport
> 
> @sebastian-philipp makes sense, let's do it then!

@jschmid1 seems like the octopus backport of this has merged, can we now update the cephadam/upgrade suite https://github.com/ceph/ceph/blob/master/qa/suites/rados/cephadm/upgrade/1-start.yaml#L3-L4 to use it?

**Comment by jschmid1** [comment]
> @jschmid1 seems like the octopus backport of this has merged, can we now update the cephadam/upgrade suite https://github.com/ceph/ceph/blob/master/qa/suites/rados/cephadm/upgrade/1-start.yaml#L3-L4 to use it?

Yes, seems right.

**Comment by neha-ojha** [comment]
> > > > DNM, needs squashing.
> > > > @neha-ojha this PR is good now.
> > > 
> > > 
> > > great, I will report back results here.
> > 
> > 
> > @sebastian-philipp I ran your build against my test branch and it is failing with the same error https://pulpito.ceph.com/nojha-2020-08-25_17:55:27-rados:cephadm:partial-upgrade-wip-swagner-testing-2020-08-25-1535-distro-basic-smithi/
> 
> Hm this can't work. We're trying to run `orch daemon redeploy --image` on a MGR that is still on 15.2.0 .
> 
> I think we have to:
> 
> 1. Merge this PR
> 2. backport to octopus
> 3. set the upgrade start to the merge commit of the backport

I did this by pointing cephadm upgrade start to https://github.com/ceph/ceph/commit/6fcc9ae18070f4d54a41e2d908d7a08d392e5194 but still see the same issue.

```
2020-09-21T20:05:51.195 INFO:teuthology.orchestra.run.smithi008:> sudo /home/ubuntu/cephtest/cephadm --image quay.ceph.io/ceph-ci/ceph:6fcc9ae18070f4d54a41e2d908d7a08d392e5194 shell -c /etc/ceph/ceph.conf -k /etc/ceph/ceph.client.admin.keyring --fsid a464be88-fc44-11ea-a2a9-001a4aab830c -e sha1=d257bf0e04b992b118804e26b1b49922d599550f -- bash -c 'ceph orch daemon redeploy --image=quay.ceph.io/ceph-ci/ceph:$sha1 --name=mgr.y'
2020-09-21T20:05:51.204 INFO:journalctl@ceph.mon.a.smithi008.stdout:Sep 21 20:05:50 smithi008 bash[9516]: cluster 2020-09-21T20:05:49.897908+0000 mgr.x (mgr.34109) 213 : cluster [DBG] pgmap v187: 1 pgs: 1 active+clean; 0 B data, 3.6 MiB used, 707 GiB / 715 GiB avail
2020-09-21T20:05:51.204 INFO:journalctl@ceph.mon.a.smithi008.stdout:Sep 21 20:05:50 smithi008 bash[9516]: audit 2020-09-21T20:05:50.560414+0000 mon.a (mon.0) 715 : audit [DBG] from='client.? 172.21.15.8:0/1686185098' entity='client.admin' cmd=[{"prefix": "health", "format": "json"}]: dispatch
2020-09-21T20:05:51.204 INFO:journalctl@ceph.mon.c.smithi008.stdout:Sep 21 20:05:50 smithi008 bash[13928]: cluster 2020-09-21T20:05:49.897908+0000 mgr.x (mgr.34109) 213 : cluster [DBG] pgmap v187: 1 pgs: 1 active+clean; 0 B data, 3.6 MiB used, 707 GiB / 715 GiB avail
2020-09-21T20:05:51.204 INFO:journalctl@ceph.mon.c.smithi008.stdout:Sep 21 20:05:50 smithi008 bash[13928]: audit 2020-09-21T20:05:50.560414+0000 mon.a (mon.0) 715 : audit [DBG] from='client.? 172.21.15.8:0/1686185098' entity='client.admin' cmd=[{"prefix": "health", "format": "json"}]: dispatch
2020-09-21T20:05:51.320 INFO:journalctl@ceph.mon.b.smithi033.stdout:Sep 21 20:05:50 smithi033 bash[9475]: cluster 2020-09-21T20:05:49.897908+0000 mgr.x (mgr.34109) 213 : cluster [DBG] pgmap v187: 1 pgs: 1 active+clean; 0 B data, 3.6 MiB used, 707 GiB / 715 GiB avail
2020-09-21T20:05:51.321 INFO:journalctl@ceph.mon.b.smithi033.stdout:Sep 21 20:05:50 smithi033 bash[9475]: audit 2020-09-21T20:05:50.560414+0000 mon.a (mon.0) 715 : audit [DBG] from='client.? 172.21.15.8:0/1686185098' entity='client.admin' cmd=[{"prefix": "health", "format": "json"}]: dispatch
2020-09-21T20:05:53.320 INFO:journalctl@ceph.mon.b.smithi033.stdout:Sep 21 20:05:52 smithi033 bash[9475]: cluster 2020-09-21T20:05:51.898493+0000 mgr.x (mgr.34109) 214 : cluster [DBG] pgmap v188: 1 pgs: 1 active+clean; 0 B data, 3.6 MiB used, 707 GiB / 715 GiB avail
2020-09-21T20:05:53.441 INFO:journalctl@ceph.mon.a.smithi008.stdout:Sep 21 20:05:52 smithi008 bash[9516]: cluster 2020-09-21T20:05:51.898493+0000 mgr.x (mgr.34109) 214 : cluster [DBG] pgmap v188: 1 pgs: 1 active+clean; 0 B data, 3.6 MiB used, 707 GiB / 715 GiB avail
2020-09-21T20:05:53.442 INFO:journalctl@ceph.mon.c.smithi008.stdout:Sep 21 20:05:52 smithi008 bash[13928]: cluster 2020-09-21T20:05:51.898493+0000 mgr.x (mgr.34109) 214 : cluster [DBG] pgmap v188: 1 pgs: 1 active+clean; 0 B data, 3.6 MiB used, 707 GiB / 715 GiB avail
2020-09-21T20:05:54.232 INFO:teuthology.orchestra.run.smithi008.stderr:[errno 13] RADOS permission denied (error connecting to the cluster)
2020-09-21T20:05:54.825 DEBUG:teuthology.orchestra.run:got remote process result: 13
```

https://pulpito.ceph.com/nojha-2020-09-21_19:45:32-rados:cephadm:partial-upgrade-master-distro-basic-smithi/

@jschmid1 Am I missing something?

**Pull request merged**